hphp/runtime/vm/jit/vasm-flags.cpp

```diff
@@ -0,0 +1,215 @@
+
+#include "hphp/util/arch.h"
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+///////////////////////////////////////////////////////////////////////////////
+
+struct Env {
+  Vunit& unit;
+
+  struct DefUses {
+    Vinstr*               def;  // The single def of a VregSF
+    jit::vector<Vinstr*>  uses; // The uses of a VregSF
+  };
+
+  // Map of VregSF to its def and uses.
+  jit::hash_map<size_t, DefUses> sf_def_uses;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+Vflags getFlags(const Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, uses, defs) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+    }
+#define Inone
+#define IF(fl) return i.fl;
+#define IC(cc)
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+void setFlags(Vinstr* inst, Vflags flags) {
+  switch (inst->op) {
+#define O(name, imms, uses, defs) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+    }
+#define Inone
+#define IF(fl) i.fl = flags; return;
+#define IC(cc)
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * For each use of an SF, return the StatusFlags bits needed by the use's CC.
+ *
+ * Specializations are below, which can additionally change the CC if the
+ * platform doesn't have a simple CC mapping for the given def and use.
+ */
+template<typename Inst>
+Vflags annotateSFUses(Env& env, Inst& inst) {
+  return static_cast<Vflags>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+Vflags annotateSFUses(Env& env, jcc& inst) {
+  if (env.sf_def_uses[inst.sf].def->op == Vinstr::ucomisd) {
+    switch (inst.cc) {
+    case CC_AE: // CC_NB
+      inst.cc = CC_GE;
+      break;
+    default:
+      break;
+    }
+  }
+  return static_cast<Vflags>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * Map this uses CC in any platform-specific way and return the platform-
+ * specific status flags bits that are used for the CC.
+ */
+Vflags annotateSFUses(Env& env, Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, ...) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+  }
+#define Inone
+#define IF(fl)
+#define IC(cc) return annotateSFUses(env, i);
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+}
+
+/*
+ * Map SF pass for a Vunit.
+ */
+void annotateSFUses(Vunit& unit) {
+
+  if (!arch_any(Arch::ARM)) return;
```

Comment:
This is unnecessary—we can just not invoke this pass for the other platforms.

Also: Don't you want to call this somewhere in `emitARM()` or `optimizeARM()`?